{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DNA To Music\n",
    "#### Author: Frank Escalante\n",
    "#### Updated: 08/10/2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.image as img\n",
    "import cv2\n",
    "import random\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DNA_nmbrs(bases, seq_length):\n",
    "    nucleotide = ''\n",
    "    base_combos = []\n",
    "    count = 0\n",
    "    while len(base_combos) != len(bases)**seq_length:\n",
    "        while len(nucleotide) < seq_length and len(nucleotide) != seq_length:\n",
    "            base = random.choice(bases)\n",
    "            nucleotide += base\n",
    "        if nucleotide not in base_combos:\n",
    "            base_combos.append((count, nucleotide))\n",
    "            nucleotide = ''\n",
    "            count += 1\n",
    "        if nucleotide in base_combos:\n",
    "            nucleotide = ''\n",
    "    return len(base_combos), base_combos\n",
    "                \n",
    "        \n",
    "def fib_seq(N,seq=[]):\n",
    "    if N == 0:\n",
    "        seq = [0]\n",
    "    elif N == 1:\n",
    "        seq = [0,1]\n",
    "    elif N > 1:\n",
    "        seq = [0,1]\n",
    "        for i in range(2,N):\n",
    "            next = seq[len(seq)-1] + seq[len(seq)-2]\n",
    "            seq.append(next)\n",
    "    return seq\n",
    "   \n",
    "        \n",
    "def mkDNA(length, alphabet):\n",
    "    if alphabet == 'STD':\n",
    "        bits = 'ATGC'\n",
    "    else:\n",
    "        bits = alphabet\n",
    "        \n",
    "    max_numb = len(bits)\n",
    "    lst = []\n",
    "    for i in range(0,max_numb):\n",
    "        lst.append(i)\n",
    "    \n",
    "    DNA = ''\n",
    "    for i in range(0,length):\n",
    "        select = random.choice(lst)\n",
    "        base = bits[select]\n",
    "        DNA += base\n",
    "    \n",
    "    return DNA\n",
    "\n",
    "def music_notes(alphabet):\n",
    "    if alphabet == 'STD':\n",
    "        alphabet = ['C','C#','D','D#','E','F','F#','G','G#','A','A#','B']\n",
    "    if alphabet == 'Expanded':\n",
    "        alphabet = ['R','C','C#','D','D#','E','E#','F','F#','G','G#','A','A#','B','B#','X']\n",
    "    music = []\n",
    "    for i in range(0, len(alphabet)):\n",
    "        music.append((i,alphabet[i]))\n",
    "    return music\n",
    "\n",
    "def quartenary(nmbr):\n",
    "    nmbr_dict = {0: \"A\", 1: \"C\", 2: \"G\", 3: \"T\"}\n",
    "    memory = []\n",
    "    nucleotides = []\n",
    "    if nmbr <= 3:\n",
    "            remainder = nmbr % 4\n",
    "            memory.append(0)\n",
    "            memory.append(0)\n",
    "            memory.append(remainder)\n",
    "    else:\n",
    "        fours = nmbr // 4\n",
    "        if fours <= 3:\n",
    "            memory.append(0)\n",
    "            memory.append(fours)\n",
    "            memory.append(nmbr % 4)\n",
    "        elif fours > 3:\n",
    "            sixteens = nmbr // 16\n",
    "            memory.append(sixteens)\n",
    "            remainder = nmbr % 16\n",
    "            if remainder == 0:\n",
    "                memory.append(0)\n",
    "                memory.append(0)\n",
    "            elif remainder > 0:\n",
    "                memory.append(round(remainder // 4))\n",
    "                memory.append(remainder % 4)\n",
    "    \n",
    "    for i in memory:\n",
    "        nucleotide = nmbr_dict[i]\n",
    "        nucleotides.append(nucleotide)\n",
    "    \n",
    "    codon = ''\n",
    "    for i in nucleotides:\n",
    "        codon += i\n",
    "        \n",
    "    return codon, memory\n",
    "\n",
    "def mltplr(numbr, points, multiplier):\n",
    "    modulus = len(points)\n",
    "    original = numbr % modulus\n",
    "    new = (numbr * multiplier) % modulus\n",
    "    if original == 0:\n",
    "        original = modulus\n",
    "    if new == 0:\n",
    "        new = modulus\n",
    "    for i in points:\n",
    "        if i[0] == original:\n",
    "            index1 = i[0]\n",
    "            DNA_codon1 = i[1]\n",
    "            music_note1 = i[2]\n",
    "        if i[0] == new:\n",
    "            index2 = i[0]\n",
    "            DNA_codon2 = i[1]\n",
    "            music_note2 = i[2]\n",
    "    seq = ''\n",
    "    for i in points:\n",
    "        if i[0] == abs(index2 - index1) % modulus:\n",
    "            seq = i[1]\n",
    "    if seq == '':\n",
    "        seq = points[modulus-1][1]\n",
    "    return (seq,(music_note1,music_note2))   \n",
    "    \n",
    "def DNAmusic(seq, points):\n",
    "    count = 0\n",
    "    codon = ''\n",
    "    number_seq = []\n",
    "    music_seq = []\n",
    "    for i in range(0,len(seq)):\n",
    "        if i % 3 != 0:\n",
    "            codon += seq[i]\n",
    "            count += 1\n",
    "        if i % 3 == 0:\n",
    "            codon += seq[i]\n",
    "            count = 0\n",
    "            for entry in points:\n",
    "                trans_seq = entry[1]\n",
    "                if trans_seq == codon:\n",
    "                    number_seq.append(entry[0])\n",
    "                    music_seq.append(entry[2])\n",
    "            codon = ''\n",
    "    return number_seq, music_seq\n",
    "\n",
    "def music_nlyzr(music,nmbr_seq):\n",
    "    unique_notes = list(set(music))\n",
    "    unique_nmbrs = list(set(nmbr_seq))\n",
    "    notes = {}\n",
    "    nmbrs = {}\n",
    "    for i in music:\n",
    "        if i != list(notes.keys()):\n",
    "            notes[i] = music.count(i)\n",
    "    for i in nmbr_seq:\n",
    "        if i != list(nmbrs.keys()):\n",
    "            nmbrs[i] = nmbr_seq.count(i)\n",
    "    return nmbrs, notes \n",
    "\n",
    "def synDNA_mltplr(factor,points,input_seq):\n",
    "    out_seq = [] \n",
    "    syn_DNA = ''\n",
    "    music = ''\n",
    "    for nmbr in input_seq:\n",
    "        out_seq.append(factor*nmbr)\n",
    "    for nmbr in out_seq:\n",
    "        for point in points:\n",
    "            if point[0] == nmbr % len(points) or nmbr % len(points) == 0:\n",
    "                codon = point[1]\n",
    "                note = point[2]\n",
    "                syn_DNA += codon\n",
    "                music += note + ','\n",
    "                break\n",
    "    music = music[:-1]\n",
    "    return [music], [syn_DNA]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "codons = []\n",
    "for i in range(0,64):\n",
    "    codons.append((i,quartenary(i)[0]))\n",
    "\n",
    "notes = music_notes('Expanded')\n",
    "\n",
    "cdn_nmbr_nte = []\n",
    "count = 0\n",
    "countr = 0\n",
    "while count != len(codons):\n",
    "    code = (count+1, codons[count][1], notes[countr][1])\n",
    "    cdn_nmbr_nte.append(code)\n",
    "    count += 1\n",
    "    countr += 1\n",
    "    if countr == 16:\n",
    "         countr = 0\n",
    "            \n",
    "notes = []\n",
    "DNAseq = ''\n",
    "for i in range(0,64):\n",
    "    output = mltplr(i,cdn_nmbr_nte,38)\n",
    "    DNAseq += output[0]\n",
    "\n",
    "    \n",
    "sny_DNA_2x = synDNA_mltplr(2,cdn_nmbr_nte,[i for i in range(1,65)])\n",
    "\n",
    "GFP = 'ATGCGTAAAGGAGAAGAACTTTTCACTGGAGTTGTCCCAATTCTTGTTGAATTAGATGGTGA\\\n",
    "TGTTAATGGGCACAAATTTTCTGTCAGTGGAGAGGGTGAAGGTGATGCAACAAACGGAAAACTTACCCTTAAATTTATT\\\n",
    "TGCACTACTGGAAAACTACCTGTTCCATGGCCAACACTTGTCACTACTTTGACTTATGGTGTTCAATGC\\\n",
    "TTTGCGAGATACCCAGATCATATGAAACAGCATGACTTTTTCAAGAGTGCCATGCCCGAAGGTTATGTAC\\\n",
    "AGGAAAGAACTATATCTTTCAAAGATGACGGGACCTACAAGACACGTGCTGAAGTCAAGTTTGAAGGTGA\\\n",
    "TACCCTTGTTAATAGAATCGAGTTAAAAGGTATTGATTTTAAAGAAGATGGAAACATTCTTGGACACAAA\\\n",
    "TTGGAATACAACTTTAACTCACACAATGTATACATCACGGCAGACAAACAAAAGAATGGAATCAAAGCTA\\\n",
    "ACTTCAAAATTAGACACAACGTTGAAGATGGAAGCGTTCAACTAGCAGACCATTATCAACAAAATACTCC\\\n",
    "AATTGGCGATGGCCCTGTCCTTTTACCAGACAACCATTACCTGTCCACACAATCTGCCCTTTCGAAAGAT\\\n",
    "CCCAACGAAAAGAGAGACCACATGGTCCTTCTTGAGTTTGTAACAGCTGCTGGGATTACACATGGCATGG\\\n",
    "ATGAACTATACAAAAGGCCTGCAGCAAACGACGAAAACTACGCTGCATCAGTT' \n",
    "\n",
    "RFP = 'ATGGTGAGCAAGGGCGAGGAGGATAACATGGCCATCATCAAGGAGTTCATGCGCTTCAAGGTGCACATGG\\\n",
    "AGGGCTCCGTGAACGGCCACGAGTTCGAGATCGAGGGCGAGGGCGAGGGCCGCCCCTACGAGGGCACCCA\\\n",
    "GACCGCCAAGCTGAAGGTGACCAAGGGTGGCCCCCTGCCCTTCGCCTGGGACATCCTGTCCCCTCAGTTC\\\n",
    "ATGTACGGCTCCAAGGCCTACGTGAAGCACCCCGCCGACATCCCCGACTACTTGAAGCTGTCCTTCCCCG\\\n",
    "AGGGCTTCAAGTGGGAGCGCGTGATGAACTTCGAGGACGGCGGCGTGGTGACCGTGACCCAGGACTCCTC\\\n",
    "CCTGCAGGACGGCGAGTTCATCTACAAGGTGAAGCTGCGCGGCACCAACTTCCCCTCCGACGGCCCCGTA\\\n",
    "ATGCAGAAGAAGACCATGGGCTGGGAGGCCTCCTCCGAGCGGATGTACCCCGAGGACGGCGCCCTGAAGG\\\n",
    "GCGAGATCAAGCAGAGGCTGAAGCTGAAGGACGGCGGCCACTACGACGCTGAGGTCAAGACCACCTACAA\\\n",
    "GGCCAAGAAGCCCGTGCAGCTGCCCGGCGCCTACAACGTCAACATCAAGTTGGACATCACCTCCCACAAC\\\n",
    "GAGGACTACACCATCGTGGAACAGTACGAACGCGCCGAGGGCCGCCACTCCACCGGCGGCATGGACGAGC\\\n",
    "TGTACAAGTAA'\n",
    "\n",
    "music_trans1 = DNAmusic(GFP,cdn_nmbr_nte)\n",
    "GFP_analysis = music_nlyzr(music_trans1[1],music_trans1[0])\n",
    "\n",
    "music_trans2 = DNAmusic(RFP,cdn_nmbr_nte)\n",
    "RFP_analysis = music_nlyzr(music_trans2[1],music_trans2[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The gene for GFP is approximately: 753 nucleotides long or: 251.0 Amino Acids Long\n",
      "The gene for RFP is approximately: 711 nucleotides long or: 237.0 Amino Acids Long\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(['C,C,D,E,G,X,G,G,D,B,C,X,C,C,D,E,G,X,G,G,D,B,C,R,C,C,D,E,G,X,G,G,D,B,C,X,C,C,D,E,G,X,G,G,D,B,C,R,C,C,D,E,G,X,G,G,D,B,C,X,C,C,D'],\n",
       " ['AACAACAATACCAGCATTCGCGGCAATGTCTACCTTCACTACAATTCCTGCGTTGGCCGCAATCTCGACAAAGACGACAATGCCGGCATTTGCAGCAATATCCACCTTTACCACAATCCCCGCGTTAGCTGCAATTTCAACAAAAACAACAATACCAGCATTCGCGGCAATGTCTACCTTCACTACAAT'])"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"The gene for GFP is approximately: \" + str(len(GFP)) +\" nucleotides long\\\n",
    " or: \" + str(len(GFP)/3) + \" Amino Acids Long\")\n",
    "print(\"The gene for RFP is approximately: \" + str(len(RFP)) +\" nucleotides long\\\n",
    " or: \" + str(len(RFP)/3) + \" Amino Acids Long\")\n",
    "\n",
    "# Generate a random nucleotide sequence 168 kbp long\n",
    "syn_DNA = mkDNA(16800,'STD')\n",
    "\n",
    "music_trans_synDNA = DNAmusic(syn_DNA,cdn_nmbr_nte)\n",
    "synDNA_analysis = music_nlyzr(music_trans_synDNA[1],music_trans_synDNA[0])\n",
    "#print(music_trans_synDNA)\n",
    "#print(synDNA_analysis)\n",
    "\n",
    "Fibonacci_seq_DNA = synDNA_mltplr(2,cdn_nmbr_nte,fib_seq(64)[1:])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Works Cited\n",
    "1) GFP Sequence\n",
    "\n",
    "“Aequorea Victoria Isolate SGFP-206 Green Fluorescent Protein (GFP) \n",
    "          Gen - Nucleotide - NCBI.” National Center for Biotechnology \n",
    "          Information, U.S. National Library of Medicine, \n",
    "          www.ncbi.nlm.nih.gov/nuccore/JX472995.1?report=fasta. \n",
    "          Accessed 10 Aug. 2024. \n",
    " \n",
    " \n",
    "2) RFP Sequence\n",
    "\n",
    "“Synthetic Construct Monomeric Red Fluorescent Protein Gene, Complete \n",
    "          C - Nucleotide - NCBI.” National Center for Biotechnology \n",
    "          Information, U.S. National Library of Medicine, \n",
    "          www.ncbi.nlm.nih.gov/nuccore/AY678264.1?report=fasta. \n",
    "          Accessed 10 Aug. 2024. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
